lib/AuthStateManager.ts

```diff
@@ -0,0 +1,148 @@
+import { AuthSdkError } from './errors';
+import { AuthState, UpdateAuthStateOptions } from './types';
+import { OktaAuth } from './browser';
+const PCancelable = require('p-cancelable');
+
+export const DEFAULT_AUTH_STATE = { 
+  isPending: true,
+  isAuthenticated: false,
+  idToken: null,
+  accessToken: null,
+};
+const DEFAULT_PENDING = {
+  updateAuthStatePromise: null,
+  canceledTimes: 0
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+const MAX_PROMISE_CANCEL_TIMES = 10;
+
+class AuthStateManager {
+  _sdk: OktaAuth;
+  _pending: { 
+    updateAuthStatePromise: typeof PCancelable;
+    canceledTimes: number; 
+  };
+  _authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this._sdk = sdk;
+    this._pending = { ...DEFAULT_PENDING };
+    this._authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to start updateState process
+    // "added" event is emitted in both add and renew process, it will be canceled by the followed "renewed" event
```

Comment:
what value is "renewed" adding? Is "added" and "removed" sufficient?

Reply:
true, it would be canceled anyway, it's added for the `devMode` to see clear events flow.

Reply:
If devMode is true maybe we can add a separate block where each event is listened to and simply logged, then the logging logic is independent from functional logic, here we need to decide what inputs will trigger creation of a new state. I think it would be simpler if we did not consider "renewed" but relied only on "added" or "removed"

Reply:
ok, we can just get `renewed` listener removed from here. If in the future we want to debug `renew` event flow, we can just add log in tokenManager. It makes sense to just log what matters to `AuthStateManager`.